tracing/traced_file.cc

```diff
@@ -0,0 +1,185 @@
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/future.hh>
+#include <seastar/core/print.hh>
+
+#include "tracing/traced_file.hh"
+
+using namespace seastar;
+
+class traced_file_impl : public file_impl {
+private:
+    // Invariant: _f_impl == *get_file_impl(_f).
+    // The _f is only stored to keep ownership.
+    file _f;
+    file_impl& _f_impl;
+
+    tracing::trace_state_ptr _trace_state;
+    sstring _trace_prefix;
+
+    template <typename... Args>
+    void trace(const char* fmt, Args&&... args) {
+        tracing::trace(_trace_state, "{} {}", _trace_prefix, format(fmt, std::forward<Args>(args)...));
+    }
+
+public:
+    traced_file_impl(file, tracing::trace_state_ptr, sstring);
+
+    virtual future<size_t> write_dma(uint64_t pos, const void* buf, size_t len, const io_priority_class& pc) override;
+    virtual future<size_t> write_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override;
+    virtual future<size_t> read_dma(uint64_t pos, void* buf, size_t len, const io_priority_class& pc) override;
+    virtual future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override;
+    virtual future<> flush(void) override;
+    virtual future<struct stat> stat(void) override;
+    virtual future<> truncate(uint64_t length) override;
+    virtual future<> discard(uint64_t offset, uint64_t length) override;
+    virtual future<> allocate(uint64_t position, uint64_t length) override;
+    virtual future<uint64_t> size(void) override;
+    virtual future<> close() override;
+    virtual std::unique_ptr<seastar::file_handle_impl> dup() override;
+    virtual subscription<directory_entry> list_directory(std::function<future<> (directory_entry de)> next) override;
+    virtual future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t range_size, const io_priority_class& pc) override;
+};
+
+traced_file_impl::traced_file_impl(file f, tracing::trace_state_ptr trace_state, sstring trace_prefix)
+    : _f(std::move(f)), _f_impl(*get_file_impl(_f)), _trace_state(std::move(trace_state)), _trace_prefix(std::move(trace_prefix)) {}
+
+future<size_t>
+traced_file_impl::write_dma(uint64_t pos, const void* buf, size_t len, const io_priority_class& pc) {
+    trace("call write_dma(pos = {}, const void* buf, len = {}, const io_priority_class&)", pos, len);
```

Comment:
Please use user-understandable terms, not function names.

It will be nice to print out the priority class in a user-understandlable way, but that can be done later.